fineftp-server/src/ftp_session.cpp

```diff
@@ -1302,7 +1304,31 @@ namespace fineftp
                                                         }
                                                         else
                                                         {
-                                                          me->sendFtpMessage(FtpReplyCode::CLOSING_DATA_CONNECTION, "Done");
+                                                          // Ugly work-around:
+                                                          // An FTP client implementation has been observed to close the data connection
+                                                          // as soon as it receives the 226 status code - even though it hasn't received
+                                                          // all data, yet. To improve interoperability with such buggy clients, sending
+                                                          // of the 226 status code can be delayed a bit. The delay is defined through a
+                                                          // preprocessor definition. If the delay is 0, no delay is introduced at all.
+                                                          #if (0 == DELAY_226_RESP_MS)
+                                                            me->sendFtpMessage(FtpReplyCode::CLOSING_DATA_CONNECTION, "Done");
+                                                          #else
+                                                            // Close Data Socket properly
+                                                            {
+                                                              asio::error_code ec;
+                                                              data_socket->shutdown(asio::socket_base::shutdown_both, ec);
+                                                              data_socket->close(ec);
+                                                            }
+
```

Comment:
@bjuulp: This must always be executed, right? I think the current implementation does not close the data socket at all, after sending a file...

My old implementation did close the socket, but the new memory-mapped file implementation does not use the `writeDataToSocket` function anymore, which would close the socket. So I think this bug was introduced with the memory-mapped file implementation. Funny, that curl doesn't complain about that.

Reply:
I've now tested what happens if I explicitly close the socket and set the delay to 0. This still triggers the bug in the old lwftp implementation. So, the only thing I've done in the latest commit is to move the explicit connection closure so that it is done irrespective of the presence or absence of the delay.